File: Antidote/AvatarManager.m
//
//  AvatarManager.m
//  Antidote
//
//  Created by Dmitry Vorobyov on 05.10.14.
//  Copyright (c) 2014 dvor. All rights reserved.
//

#import "AvatarManager.h"
#import "NSString+Utilities.h"
#import "ToxManager.h"
#import "ProfileManager.h"

@interface AvatarManager()

@property (strong, nonatomic) NSCache *avatarCache;

@end

@implementation AvatarManager

#pragma mark -  Lifecycle

- (id)init
{
    return nil;
}

- (id)initPrivate
{
    if (self = [super init]) {
        self.avatarCache = [NSCache new];

        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(updateFriendNotification:)
                                                     name:kToxFriendsContainerUpdateSpecificFriendNotification
                                                   object:nil];
    }

    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

+ (AvatarManager *)sharedInstance
{
    static AvatarManager *instance;
    static dispatch_once_t onceToken;

    dispatch_once(&onceToken, ^{
        instance = [[AvatarManager alloc] initPrivate];
    });

    return instance;
}

#pragma mark -  Public

+ (UIImage *)avatarInCurrentProfileWithClientId:(NSString *)clientId
                       orCreateAvatarFromString:(NSString *)string
                                       withSide:(CGFloat)side
{
    if (clientId) {
        UIImage *image = [[self sharedInstance].avatarCache objectForKey:clientId];

        if (image) {
            return image;
        }

        NSString *path = [[ProfileManager sharedInstance] pathInAvatarDirectoryForFileName:clientId];

        if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
            image = [UIImage imageWithContentsOfFile:path];

            if (image) {
                [[self sharedInstance].avatarCache setObject:image forKey:clientId];
                return image;
            }
        }
    }

    return [self avatarFromString:string side:side];
}

+ (UIImage *)avatarFromString:(NSString *)string side:(CGFloat)side
{
    string = [self firstLettersOfTwoWordsLettersFromString:string];

    UILabel *label = [UILabel new];
    label.backgroundColor = [AppearanceManager bubbleOutgoingColor];
    label.layer.borderColor = [AppearanceManager textMainColor].CGColor;
    label.layer.borderWidth = 1.0;
    label.layer.masksToBounds = YES;
    label.textColor = [AppearanceManager textMainColor];
    label.textAlignment = NSTextAlignmentCenter;
    label.text = string;

    CGFloat fontSize = 51;
    CGSize size;

    do {
        fontSize--;

        // PL - placeholder in case if avatar text is nil
        NSString *str = string ?: @"PL";

        size = [str stringSizeWithFont:[AppearanceManager fontHelveticaNeueLightWithSize:fontSize]];

    } while (MAX(size.width, size.height) > side);

    CGRect frame = CGRectZero;
    frame.size.width = frame.size.height = side;

    label.font = [AppearanceManager fontHelveticaNeueLightWithSize:(int) (fontSize * 0.6)];
    label.layer.cornerRadius = frame.size.width / 2;
    label.frame = frame;

    return [self imageWithView:label];
}

+ (void)clearCache
{
    [[self sharedInstance].avatarCache removeAllObjects];
}

#pragma mark -  Notifications

- (void)updateFriendNotification:(NSNotification *)notification
{
    ToxFriend *friend = notification.userInfo[kToxFriendsContainerUpdateKeyFriend];

    if (friend.clientId) {
        [self.avatarCache removeObjectForKey:friend.clientId];
    }
}

#pragma mark -  Private

+ (NSString *)firstLettersOfTwoWordsLettersFromString:(NSString *)string
{
    NSArray *array = [string componentsSeparatedByCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];

    NSString *result = nil;

    for (NSUInteger index = 0; index < array.count; index++) {
        NSString *word = array[index];

        if (! word.length) {
            continue;
        }

        NSString *firstLetter = [word substringToIndex:1];

        if (result.length == 2) {
            break;
        }
        else if (result.length == 1) {
            result = [result stringByAppendingString:firstLetter];
        }
        else {
            result = firstLetter;
        }
    }

    return [result uppercaseString];
}

+ (UIImage *)imageWithView:(UIView *)view
{
    UIGraphicsBeginImageContextWithOptions(view.bounds.size, view.opaque, 0.0);
    [view.layer renderInContext:UIGraphicsGetCurrentContext()];

    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();

    UIGraphicsEndImageContext();

    return image;
}

@end

File: Antidote/AvatarManager.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import Foundation

class AvatarManager {
    enum AvatarType: String {
        case Normal
        case Call
    }

    fileprivate let theme: Theme
    fileprivate let cache: NSCache<AnyObject, AnyObject>

    init(theme: Theme) {
        self.theme = theme
        self.cache = NSCache()
    }

    /**
        Returns round avatar created from string with a given diameter. Searches for an avatar in cache first,
        if not found creates it.

        - Parameters:
          - string: String to create avatar from. In case of empty string avatar will be set to "?".
          - diameter: Diameter of circle with avatar.

        - Returns: Avatar from given string with given size.
     */
    func avatarFromString(_ string: String, diameter: CGFloat, type: AvatarType = .Normal) -> UIImage {
        var string = string

        if string.isEmpty {
            string = "?"
        }

        let key = keyFromString(string, diameter: diameter, type: type)

        if let avatar = cache.object(forKey: key as AnyObject) as? UIImage {
            return avatar
        }

        let avatar = createAvatarFromString(string, diameter: diameter, type: type)
        cache.setObject(avatar, forKey: key as AnyObject)

        return avatar
    }
}

private extension AvatarManager {
    func keyFromString(_ string: String, diameter: CGFloat, type: AvatarType) -> String {
        return "\(string)-\(diameter)-\(type.rawValue)"
    }

    func createAvatarFromString(_ string: String, diameter: CGFloat, type: AvatarType) -> UIImage {
        let avatarString = avatarStringFromString(string)

        let label = UILabel()
        label.layer.borderWidth = 1.0
        label.layer.masksToBounds = true
        label.textAlignment = .center
        label.text = avatarString

        switch type {
            case .Normal:
                label.backgroundColor = theme.colorForType(.NormalBackground)
                label.layer.borderColor = theme.colorForType(.LinkText).cgColor
                label.textColor = theme.colorForType(.LinkText)
            case .Call:
                label.backgroundColor = .clear
                label.layer.borderColor = theme.colorForType(.CallButtonIconColor).cgColor
                label.textColor = theme.colorForType(.CallButtonIconColor)
        }

        var size: CGSize
        var fontSize = diameter

        repeat {
            fontSize -= 1

            let font = UIFont.antidoteFontWithSize(fontSize, weight: .light)
            size = avatarString.stringSizeWithFont(font)
        }
        while (max(size.width, size.height) > diameter)

        let frame = CGRect(x: 0, y: 0, width: diameter, height: diameter)

        label.font = UIFont.antidoteFontWithSize(fontSize * 0.6, weight: .light)
        label.layer.cornerRadius = frame.size.width / 2
        label.frame = frame

        return imageWithLabel(label)
    }

    func avatarStringFromString(_ string: String) -> String {
        guard !string.isEmpty else {
            return ""
        }

        // Avatar can have alphanumeric symbols and ? sign.
        let badSymbols = (CharacterSet.alphanumerics.inverted as NSCharacterSet).mutableCopy() as! NSMutableCharacterSet
        badSymbols.removeCharacters(in: "?")

        let words = string.components(separatedBy: CharacterSet.whitespaces).map {
            $0.components(separatedBy: badSymbols as CharacterSet).joined(separator: "")
        }.filter {
            !$0.isEmpty
        }

        let result = words.map {
            $0.isEmpty ? "" : $0[0..<1]
        }.joined(separator: "")

        let numberOfLetters = min(2, result.characters.count)

        return result.uppercased()[0..<numberOfLetters]
    }

    func imageWithLabel(_ label: UILabel) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(label.bounds.size, false, 0.0)
        label.layer.render(in: UIGraphicsGetCurrentContext()!)

        let image = UIGraphicsGetImageFromCurrentImageContext()

        UIGraphicsEndImageContext()

        return image!
    }
}
